forc-pkg/src/pkg.rs

```diff
@@ -2090,12 +2090,92 @@ pub const SWAY_BIN_HASH_SUFFIX: &str = "-bin-hash";
 pub const SWAY_BIN_ROOT_SUFFIX: &str = "-bin-root";
 
 /// Builds a project with given BuildOptions
-pub fn build_with_options(build_options: BuildOptions) -> Result<Compiled> {
+pub fn build_with_options(build_options: BuildOptions) -> Result<()> {
+    let path = &build_options.path;
+
+    let this_dir = if let Some(ref path) = path {
+        PathBuf::from(path)
+    } else {
+        std::env::current_dir()?
+    };
+
+    // Decide the type of the ManifestFile at this_dir
+    if let Ok(package_manifest) = PackageManifestFile::from_dir(&this_dir) {
+        build_package_with_options(&package_manifest, build_options)?;
+    } else if let Ok(workspace_manifest) = WorkspaceManifestFile::from_dir(&this_dir) {
+        build_workspace_with_options(&workspace_manifest, build_options)?;
+    } else {
+        bail!(format!(
+            "Cannot find {} {:?}",
+            sway_utils::MANIFEST_FILE_NAME,
+            this_dir
+        ))
+    }
+    Ok(())
+}
+
+fn build_workspace_with_options(
+    manifest: &WorkspaceManifestFile,
+    build_options: BuildOptions,
+) -> Result<()> {
+    let workspace_dir = PathBuf::from(manifest.dir());
+    // Find root member of the workspace and build it.
+    let root_member = find_root_workspace_member(manifest, &build_options)?;
+    let root_member_path = workspace_dir.join(root_member);
+    let root_member_manifest = PackageManifestFile::from_dir(&root_member_path)?;
+    build_package_with_options(&root_member_manifest, build_options)?;
+    Ok(())
+}
+
+fn find_root_workspace_member(
+    manifest: &WorkspaceManifestFile,
+    build_options: &BuildOptions,
+) -> Result<String> {
+    let locked = build_options.locked;
+    let offline = build_options.offline_mode;
+    let members: HashSet<_> = manifest.members().cloned().collect();
+    for (member, member_path) in manifest.members().zip(manifest.member_paths()?) {
+        info!("Checking {member}");
+        let member_pkg_manifest = PackageManifestFile::from_dir(member_path.as_path())?;
+        let build_plan = BuildPlan::from_lock_and_manifest(&member_pkg_manifest, locked, offline)?;
+        let graph = build_plan.graph();
+        let discovered_deps = discover_contract_dependencies(graph, &members, member.clone())?;
+        // If we can find all members of the workspace from this member, this member is the root of the workspace.
+        // Meaning that building/deploying this member will ensure whole workspace will be handled in the
```

Comment:
By stating "this member", I am referring to the graph of this member. I will improve the wordings here